test-util/src/main/java/io/zeebe/test/util/bpmn/random/blocks/EventBasedGatewayBlockBuilder.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.zeebe.test.util.bpmn.random.blocks;
+
+import io.zeebe.model.bpmn.builder.AbstractFlowNodeBuilder;
+import io.zeebe.test.util.bpmn.random.BlockBuilder;
+import io.zeebe.test.util.bpmn.random.BlockBuilderFactory;
+import io.zeebe.test.util.bpmn.random.ConstructionContext;
+import io.zeebe.test.util.bpmn.random.ExecutionPathSegment;
+import io.zeebe.test.util.bpmn.random.IDGenerator;
+import io.zeebe.test.util.bpmn.random.steps.StepPublishMessage;
+import io.zeebe.util.collection.Tuple;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+
+public class EventBasedGatewayBlockBuilder implements BlockBuilder {
+
+  private static final String CORRELATION_KEY_FIELD = "correlationKey";
+  private static final String CORRELATION_KEY_VALUE = "default_correlation_key";
+  private static final String MESSAGE_NAME_PREFIX = "message_";
+
+  private final String forkGatewayId;
+  private final String joinGatewayId;
+  private final List<Tuple<String, BlockBuilder>> branches = new ArrayList<>();
+
+  public EventBasedGatewayBlockBuilder(final ConstructionContext context) {
+    final Random random = context.getRandom();
+    final IDGenerator idGenerator = context.getIdGenerator();
+    final int maxBranches = context.getMaxBranches();
+
+    forkGatewayId = "fork_" + idGenerator.nextId();
+    joinGatewayId = "join_" + idGenerator.nextId();
+
+    final var blockSequenceBuilderFactory = context.getBlockSequenceBuilderFactory();
+
+    // must have at least two branches
+    final int numberOfBranches = Math.max(2, random.nextInt(maxBranches));
+
+    for (int i = 0; i < numberOfBranches; i++) {
+      final var branchId = idGenerator.nextId();
+      final var branchBuilder =
+          blockSequenceBuilderFactory.createBlockSequenceBuilder(context.withIncrementedDepth());
```

Comment:
Are you sure the depth should be incremented here? If so I think you should also check the currentDepth and the maxDepth and whether you're allowed to go deeper. (See SubProcessBlockBuilder for example). Personally, I think this does not increase the depth. In that case also remember to disable depth in the Factory.

Reply:
I'm not sure :sweat_smile: I looked at the parallel and exclusive gateway and did the same here. 

@pihme does a gateway adding depth? and why?

Reply:
It does add depth - in a way. It's more a technical reason than a model reason.

Think of depth here as layers in a call stack and not nesting in the BPMN model.

If the gateway wouldn't add depth, you could generate infinite processes: A process which contains a gateway which on its default branch contains a gateway, which on its default branch contains a gateway ...

Maybe this issue is now less pronounced - now that we have more elements which can be added to a process. But when I fist wrote it and there were only a couple of elements, the processes exploded in complexity.